packages/pyright-internal/src/tests/samples/constrainedTypeVar19.py

```diff
@@ -0,0 +1,18 @@
+# This tests the handling of a constrained TypeVar with literal types
+# in the constraints.
+
+from typing import TypeVar, Literal, Generic
+
+T = TypeVar("T", Literal[True], Literal[False])
+
+
+class A(Generic[T]):
+    def __init__(self, null: T = False) -> None:
+        pass
+
+
+A(null=bool())  # Type error
+
+reveal_type(A(null=False), expected_text="A[Literal[False]]")
+reveal_type(A(), expected_text="A[Literal[False]]")
```

Comment:
May I ask: is `pyright` making use of the default value of the `__init__` signature in this case? If so, does it means this removes the need for a `TypeVar` default?

Reply:
Defaults for `TypeVar` are not yet officially supported. They're proposed in PEP 696, and pyright has provisional support for them.

I don't see a need for a default `TypeVar` value in this case. Pyright supports the use of a default argument value associated with a parameter whose type is `T`. Mypy doesn't support this, however, so if you want your code to work across type checkers, you might want to consider other options like overloads.